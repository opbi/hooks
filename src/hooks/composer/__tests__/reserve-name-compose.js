import reserveNameCompose from '../reserve-name-compose';

import errorMute from '../../error-mute';
import eventLogger from '../../event-logger';

const logger = {
  info: jest.fn(),
  error: jest.fn(),
};

const callOrder = jest.fn();

const createMockDecorator =
  (decoratorName) =>
  (inputFunction) =>
  (...args) => {
    callOrder(decoratorName);
    return inputFunction(...args);
  };

const decorator = createMockDecorator('decorator0');
const decorator1 = createMockDecorator('decorator1');

describe('reserveNameCompose', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('reserves the target function name if the whole chain of decorators are enhanced', () => {
    const original = () => {};
    const decorated = reserveNameCompose(decorator, decorator1)(original);
    expect(decorated.name).toBe('original');
  });

  it('essential for any decorator in the chain to access targetFunction.name if it is not on the first place', async () => {
    const original = () => {
      const e = { message: 'error' };
      throw e;
    };
    const decorated = reserveNameCompose(eventLogger(), errorMute())(original);

    try {
      const result = await decorated({}, {}, { logger });
      expect(result.message).toBe('error');
      expect(logger.info.mock.calls).toMatchSnapshot();
      expect(logger.error.mock.calls).toMatchSnapshot();
    } catch (e) {
      if (e) {
        throw Error('error is expected to be suppressed');
      }
    }
  });
});
