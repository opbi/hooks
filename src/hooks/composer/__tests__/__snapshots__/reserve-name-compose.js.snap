// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reserveNameCompose essential for any decorator in the chain to access targetFunction.name if it is not on the first place 1`] = `
Array [
  Array [
    Object {
      "event": "original",
    },
    "original",
  ],
]
`;

exports[`reserveNameCompose essential for any decorator in the chain to access targetFunction.name if it is not on the first place 2`] = `Array []`;
